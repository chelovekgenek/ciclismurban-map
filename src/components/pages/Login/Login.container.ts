import { connect } from "react-redux"
import { withFormik } from "formik"

import { validateFormik } from "helpers"
import { UserModel, ExposeGroup, LoginForm } from "models/user"
import { LoginActions, getFetching, getError } from "store/entities/user"
import { TAppState } from "store/entities"

import { IProps, Login } from "./Login"

export interface IStateProps {
  fetching: ReturnType<typeof getFetching>
  responseError: ReturnType<typeof getError>
}
export interface IDispatchProps {
  login: typeof LoginActions.request
}

export const LoginContainer = connect<IStateProps, IDispatchProps, {}, TAppState>(
  state => ({
    fetching: getFetching(state),
    responseError: getError(state),
  }),
  {
    login: LoginActions.request,
  },
)(
  withFormik<IProps, LoginForm>({
    mapPropsToValues: () => new LoginForm(),
    handleSubmit: (values, { props }) => props.login(values),
    validate: validateFormik(UserModel, [ExposeGroup.LOGIN]),
    validateOnBlur: true,
  })(Login),
)
