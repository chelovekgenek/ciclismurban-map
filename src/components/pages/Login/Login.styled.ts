import styled, { css } from "styled-components"

import * as UI from "components/generic/ui"

export const Icon = styled(UI.Icon)`
  color: rgba(0, 0, 0, 0.25);
`
interface IButtonProps {
  marginBottom?: boolean
}
export const Button = styled(UI.Button)<IButtonProps>`
  width: 100%;
  ${({ marginBottom = false }) =>
    marginBottom &&
    css`
      margin-bottom: 1.2rem;
    `}
`

export const CenterText = styled.div`
  text-align: center;
`
