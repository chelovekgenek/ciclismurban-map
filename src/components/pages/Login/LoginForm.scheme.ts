export class LoginForm {
  email = ""
  password = ""
}
