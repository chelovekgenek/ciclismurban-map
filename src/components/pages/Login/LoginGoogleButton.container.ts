import { connect } from "react-redux"

import { LoginGoogleButton } from "./LoginGoogleButton"
import { Actions } from "store/entities/auth"
import { TAppState } from "store/entities"

export interface IDispatchProps {
  login: typeof Actions.LoginByGoogle.request
}

export const LoginGoogleButtonContainer = connect<{}, IDispatchProps, {}, TAppState>(
  null,
  {
    login: Actions.LoginByGoogle.request,
  },
)(LoginGoogleButton)
