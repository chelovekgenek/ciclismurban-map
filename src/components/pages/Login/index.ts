export { LoginContainer as Login } from "./Login.container"
