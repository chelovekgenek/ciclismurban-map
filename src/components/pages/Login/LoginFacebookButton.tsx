import React, { useState, useCallback } from "react"

import FacebookLogin from "react-facebook-login/dist/facebook-login-render-props"

import { IDispatchProps } from "./LoginFacebookButton.container"
import * as LoginStyles from "./Login.styled"

interface IProps extends IDispatchProps {}

export const LoginFacebookButton: React.FC<IProps> = ({ login }) => {
  const [disabled, setDisabled] = useState(false)
  const onSucess = useCallback(r => r.accessToken && login(r.accessToken), [login])
  const handleFailure = useCallback(() => setDisabled(true), [])
  return (
    <FacebookLogin
      appId={process.env.REACT_APP_FACEBOOK_OAUTH_KEY as string}
      autoLoad
      render={(renderProps: any) => (
        <LoginStyles.Button icon="facebook" onClick={renderProps.onClick} disabled={renderProps.disabled || disabled}>
          Facebook
        </LoginStyles.Button>
      )}
      callback={onSucess}
      onFailure={handleFailure}
    />
  )
}
