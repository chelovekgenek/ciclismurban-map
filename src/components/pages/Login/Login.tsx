import React, { useEffect } from "react"
import { Link } from "react-router-dom"
import { FormikProps } from "formik"

import { Input } from "components/generic/form"
import { Divider } from "components/generic/ui"
import * as Layout from "components/generic/layout"
import { LoginForm } from "models/user"

import * as Styled from "./Login.styled"
import { IStateProps, IDispatchProps } from "./Login.container"
import { LoginGoogleButtonContainer as LoginGoogleButton } from "./LoginGoogleButton.container"
import { LoginFacebookButtonContainer as LoginFacebookButton } from "./LoginFacebookButton.container"

export interface IProps extends IStateProps, IDispatchProps {}

export const Login: React.FC<IProps & FormikProps<LoginForm>> = ({
  responseError,
  fetching,
  handleSubmit,
  isValid,
  setErrors,
}) => {
  useEffect(() => {
    if (responseError === 401) {
      setErrors(new LoginForm())
    }
  }, [responseError, setErrors])
  return (
    <Layout.Form onSubmit={handleSubmit}>
      <h2>Вход</h2>
      <Input name="email" type="email" placeholder="Email" prefix={<Styled.Icon type="user" />} />
      <Input name="password" type="password" placeholder="Password" prefix={<Styled.Icon type="lock" />} />
      <Styled.Button type="primary" htmlType="submit" disabled={!isValid} loading={fetching} marginBottom>
        Войти
      </Styled.Button>
      <Layout.Grid.Row type="flex" justify="space-between">
        <Layout.Grid.Col span={11}>
          <LoginGoogleButton />
        </Layout.Grid.Col>
        <Layout.Grid.Col span={11}>
          <LoginFacebookButton />
        </Layout.Grid.Col>
      </Layout.Grid.Row>
      <Divider>или</Divider>
      <Styled.CenterText>
        <Link to="/register">зарегестрироваться</Link>
      </Styled.CenterText>
    </Layout.Form>
  )
}
