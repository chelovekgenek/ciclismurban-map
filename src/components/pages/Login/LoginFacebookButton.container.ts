import { connect } from "react-redux"

import { Actions } from "store/entities/auth"
import { TAppState } from "store/entities"

import { LoginFacebookButton } from "./LoginFacebookButton"

export interface IDispatchProps {
  login: typeof Actions.LoginByFacebook.request
}

export const LoginFacebookButtonContainer = connect<{}, IDispatchProps, {}, TAppState>(
  null,
  {
    login: Actions.LoginByFacebook.request,
  },
)(LoginFacebookButton)
