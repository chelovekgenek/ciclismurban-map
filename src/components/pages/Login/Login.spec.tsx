import React from "react"
import { shallow } from "enzyme"

import formProps from "mocks/formik-form-props"
import { LoginForm } from "models/user"

import { Login } from "./Login"

describe("pages/Login", () => {
  const initialValues = new LoginForm()
  const render = (props: Partial<React.ComponentProps<typeof Login>> = {}) =>
    shallow(
      <Login
        {...formProps}
        initialValues={initialValues}
        values={initialValues}
        fetching={false}
        login={jest.fn() as any}
        responseError={undefined}
        {...props}
      />,
    )
  it("should match snapshot", () => {
    expect(render()).toMatchSnapshot()
  })
})
