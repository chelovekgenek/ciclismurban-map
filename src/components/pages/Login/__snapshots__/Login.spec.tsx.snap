// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Login should match snapshot 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <h2>
    Вход
  </h2>
  <Input
    name="email"
    placeholder="Email"
    prefix={
      <ForwardRef(Styled(Icon))
        type="user"
      />
    }
    type="email"
  />
  <Input
    name="password"
    placeholder="Password"
    prefix={
      <ForwardRef(Styled(Icon))
        type="lock"
      />
    }
    type="password"
  />
  <Styled(Button)
    disabled={true}
    htmlType="submit"
    loading={false}
    marginBottom={true}
    type="primary"
  >
    Войти
  </Styled(Button)>
  <Row
    gutter={0}
    justify="space-between"
    type="flex"
  >
    <Col
      span={11}
    >
      <Connect(LoginGoogleButton) />
    </Col>
    <Col
      span={11}
    >
      <Connect(LoginFacebookButton) />
    </Col>
  </Row>
  <Divider>
    или
  </Divider>
  <styled.div>
    <Link
      to="/register"
    >
      зарегестрироваться
    </Link>
  </styled.div>
</Form>
`;
